File: src/types/buisness_connection/business_connection.dto.ts

interface BusinessSummaryResponse {
  id: number;
  name: string;
  business_type: string;
}

interface UserSummaryResponse {
  id: number;
  first_name: string;
  last_name: string | null;
  login_email: string;
}

/**
 * Detailed representation of a single business connection record.
 * Used for detailed responses after creation or retrieval by ID.
 */
export interface BusinessConnectionResponse {
  /** The unique identifier of the connection record. */
  id: number;
  /** The ID of the business that initiated the connection request. */
  initiating_business_id: number;
  /** The ID of the business that received the connection request. */
  receiving_business_id: number;
  /** The defined type of business relationship. */
  connection_type:
    | "Partnership"
    | "Supplier"
    | "Client"
    | "Referral"
    | "Collaboration";
  /** The current status of the connection (e.g., 'pending', 'active'). */
  status: "pending" | "active" | "rejected" | "inactive";
  /** The ID of the user who performed the initiation (for audit). */
  initiated_by_user_id: number;
  /** Notes or message associated with the connection request or status change. */
  notes: string | null;
  /** The timestamp when the connection record was created. */
  created_at: Date;
  /** The timestamp of the last update to the connection record. */
  updated_at: Date;

 
  /** Summary details of the initiating business. */
  initiating_business: BusinessSummaryResponse;
  /** Summary details of the receiving business. */
  receiving_business: BusinessSummaryResponse;
  /** Summary details of the user who initiated the connection. */
  initiated_by_user: UserSummaryResponse;
}

/**
 * Light-weight representation for listing or summary views of business connections.
 */
export interface ConnectionSummaryResponse {
  /** The unique identifier of the connection record. */
  id: number;
  /** The ID of the initiating business. */
  initiating_business_id: number;
  /** The ID of the receiving business. */
  receiving_business_id: number;
  /** The defined type of business relationship. */
  connection_type: string;
  /** The current status of the connection. */
  status: string;
  /** The timestamp when the connection record was created. */
  created_at: Date;
}

File: internal/ports/business_connection.dto.go
package ports

import (
	"time"

	"github.com/TIA-PARTNERS-GROUP/tia-api/internal/models"
)

type CreateBusinessConnectionInput struct {
	InitiatingBusinessID uint                          `json:"initiating_business_id" validate:"required"`
	ReceivingBusinessID  uint                          `json:"receiving_business_id" validate:"required"`
	ConnectionType       models.BusinessConnectionType `json:"connection_type" validate:"required,oneof=Partnership Supplier Client Referral Collaboration"`
	InitiatedByUserID    uint                          `json:"initiated_by_user_id" validate:"required"`
	Notes                *string                       `json:"notes,omitempty"`
}

type UpdateBusinessConnectionInput struct {
	ConnectionType *models.BusinessConnectionType   `json:"connection_type,omitempty" validate:"omitempty,oneof=Partnership Supplier Client Referral Collaboration"`
	Status         *models.BusinessConnectionStatus `json:"status,omitempty" validate:"omitempty,oneof=pending active rejected inactive"`
	Notes          *string                          `json:"notes,omitempty"`
}

type BusinessConnectionResponse struct {
	ID                   uint                            `json:"id"`
	InitiatingBusinessID uint                            `json:"initiating_business_id"`
	ReceivingBusinessID  uint                            `json:"receiving_business_id"`
	ConnectionType       models.BusinessConnectionType   `json:"connection_type"`
	Status               models.BusinessConnectionStatus `json:"status"`
	InitiatedByUserID    uint                            `json:"initiated_by_user_id"`
	Notes                *string                         `json:"notes,omitempty"`
	CreatedAt            time.Time                       `json:"created_at"`
	UpdatedAt            time.Time                       `json:"updated_at"`

	// Relationships
	InitiatingBusiness BusinessResponse `json:"initiating_business"`
	ReceivingBusiness  BusinessResponse `json:"receiving_business"`
	InitiatedByUser    UserResponse     `json:"initiated_by_user"`
}

type BusinessConnectionsResponse struct {
	Connections []BusinessConnectionResponse `json:"connections"`
	Count       int                          `json:"count"`
}

func MapToBusinessConnectionResponse(bc *models.BusinessConnection) BusinessConnectionResponse {
	return BusinessConnectionResponse{
		ID:                   bc.ID,
		InitiatingBusinessID: bc.InitiatingBusinessID,
		ReceivingBusinessID:  bc.ReceivingBusinessID,
		ConnectionType:       bc.ConnectionType,
		Status:               bc.Status,
		InitiatedByUserID:    bc.InitiatedByUserID,
		Notes:                bc.Notes,
		CreatedAt:            bc.CreatedAt,
		UpdatedAt:            bc.UpdatedAt,
		InitiatingBusiness:   MapBusinessToResponse(&bc.InitiatingBusiness),
		ReceivingBusiness:    MapBusinessToResponse(&bc.ReceivingBusiness),
		InitiatedByUser:      MapUserToResponse(&bc.InitiatedByUser),
	}
}

func MapToBusinessConnectionsResponse(connections []models.BusinessConnection) BusinessConnectionsResponse {
	conns := make([]BusinessConnectionResponse, len(connections))
	for i, connection := range connections {
		conns[i] = MapToBusinessConnectionResponse(&connection)
	}

	return BusinessConnectionsResponse{
		Connections: conns,
		Count:       len(conns),
	}
}
